### Work 4_2

#### 1. Cho ví dụ về 03 đối tượng trong thực tế có thể lưu trữ dưới dạng dictionary

In [3]:
# Ví dụ 1: Thông tin sinh viên
student = {"id": "SV001", "name": "Nguyễn Văn A", "age": 20, "major": "CNTT"}

# Ví dụ 2: Bảng điểm môn học
grades = {"Toán": 9, "Lý": 8, "Hóa": 7}

# Ví dụ 3: Sản phẩm trong cửa hàng
product = {"id": "P001", "name": "Laptop", "price": 15000000, "stock": 12}

print(student)
print(grades)
print(product)

{'id': 'SV001', 'name': 'Nguyễn Văn A', 'age': 20, 'major': 'CNTT'}
{'Toán': 9, 'Lý': 8, 'Hóa': 7}
{'id': 'P001', 'name': 'Laptop', 'price': 15000000, 'stock': 12}


#### 2. Giả sử từ điển d = {'b':200, 'a':100, 'c':1}. Cho biết kết quả các câu lệnh sau
* a) d['a']
* b) d.get('e', None)
* c) len(d)
* d) d.keys()
* e) d.values()
* f) d.pop('b')
* h) giá trị của d sau khi thực hiện lệnh pop

In [4]:
d = {'b':200, 'a':100, 'c':1}

# a) d['a']
print(d['a'])  # 100

# b) d.get('e', None)
print(d.get('e', None))  # None

# c) len(d)
print(len(d))  # 3

# d) d.keys()
print(d.keys())  # dict_keys(['b', 'a', 'c'])

# e) d.values()
print(d.values())  # dict_values([200, 100, 1])

# f) d.pop('b')
print(d.pop('b'))  # 200

# h) Giá trị của d sau khi pop
print(d)  # {'a': 100, 'c': 1}


100
None
3
dict_keys(['b', 'a', 'c'])
dict_values([200, 100, 1])
200
{'a': 100, 'c': 1}


#### 3. Giả sử từ điển d = {'b':200, 'a':100, 'c':1}. Thực hiện các câu lệnh theo yêu cầu sau:
* a) Thay thế giá trị của khóa b thành số âm tương ứng.
* b) Thêm một khóa 'e' có giá trị là 500 vào d 'e':500
* c) Xóa khóa b ra khỏi từ điển (theo cách an toàn).
* d) In các cặp (key,value) ra màn hình theo thứ tự từ điển.

In [6]:
d = {'b':200, 'a':100, 'c':1}

# a) Thay thế giá trị của khóa b thành số âm
d['b'] = -d['b']
print(d)  # {'b': -200, 'a': 100, 'c': 1}

# b) Thêm khóa 'e' = 500
d['e'] = 500
print(d)  # {'b': -200, 'a': 100, 'c': 1, 'e': 500}

# c) Xóa khóa 'b' (cách an toàn: dùng pop với mặc định)
d.pop('b', None)
print(d)  # {'a': 100, 'c': 1, 'e': 500}

# d) In ra cặp (key,value) theo thứ tự từ điển (sorted key)
for k in sorted(d.keys()):
    print(k, d[k])


{'b': -200, 'a': 100, 'c': 1}
{'b': -200, 'a': 100, 'c': 1, 'e': 500}
{'a': 100, 'c': 1, 'e': 500}
a 100
c 1
e 500


#### 1. Tạo một tập hợp gồm các phần tử từ 0 đến 200. In tập hợp ra màn hình.

In [7]:
s = set(range(201))
print(s)


{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200}


#### 2. Tạo một tập hợp gồm các số nguyên tố trong khoảng (10,2000). In tập hợp ra màn hình.

In [8]:
def is_prime(n):
    if n < 2: return False
    for i in range(2, int(n**0.5)+1):
        if n % i == 0:
            return False
    return True

primes = {x for x in range(10,2001) if is_prime(x)}
print(primes)


{1031, 1543, 521, 1033, 11, 523, 13, 1549, 1039, 17, 1021, 19, 1553, 23, 1531, 1049, 1559, 1051, 29, 541, 31, 1567, 547, 1571, 37, 1061, 1063, 41, 43, 1579, 557, 1069, 47, 1583, 563, 53, 569, 59, 571, 61, 1597, 1087, 577, 1601, 67, 1091, 1093, 71, 1607, 73, 1097, 587, 1609, 1613, 79, 1103, 593, 83, 1619, 1109, 1621, 599, 89, 601, 1627, 1117, 607, 97, 1123, 101, 613, 103, 1637, 617, 1129, 107, 619, 109, 113, 631, 1657, 127, 1151, 641, 1153, 131, 643, 1663, 1667, 647, 1669, 137, 139, 1163, 653, 659, 1171, 149, 661, 151, 157, 1181, 1693, 673, 1697, 163, 1187, 677, 1699, 167, 1193, 683, 173, 1709, 1201, 179, 691, 181, 1721, 1723, 701, 1213, 191, 193, 1217, 197, 709, 199, 1223, 1733, 1229, 1741, 719, 1231, 211, 1747, 1237, 727, 1753, 733, 223, 1759, 1249, 227, 739, 229, 743, 233, 1259, 239, 751, 241, 1777, 757, 1783, 761, 251, 1787, 1277, 1789, 1279, 257, 769, 1283, 773, 263, 1289, 1801, 1291, 269, 271, 1297, 787, 1811, 277, 1301, 1303, 281, 283, 1307, 797, 1823, 293, 1319, 1831, 809, 1321,

#### 3. Phát sinh ngẫu nhiên 2 tập hợp chứa các số nguyên trong khoảng (10,2000). (Số lượng phần tử của hai tập hợp có thể khác nhau). Tìm phần giao, hội, different và symmetric_different của hai tập hợp.
* Hint: dùng import random ; random.randint

In [9]:
import random

set1 = {random.randint(10,2000) for _ in range(10)}
set2 = {random.randint(10,2000) for _ in range(15)}

print("Set1:", set1)
print("Set2:", set2)

print("Giao:", set1 & set2)
print("Hội:", set1 | set2)
print("Hiệu (set1 - set2):", set1 - set2)
print("Đối xứng:", set1 ^ set2)

Set1: {392, 1738, 747, 815, 1171, 1813, 438, 823, 665, 635}
Set2: {96, 1184, 1410, 312, 551, 1439, 1512, 1745, 49, 23, 535, 1784, 1434, 1915, 1695}
Giao: set()
Hội: {1410, 392, 1738, 1784, 1745, 1171, 1813, 23, 535, 665, 1434, 1439, 96, 1184, 1695, 1915, 551, 1512, 747, 815, 49, 438, 823, 312, 635}
Hiệu (set1 - set2): {392, 1738, 747, 815, 1171, 1813, 438, 823, 665, 635}
Đối xứng: {1410, 392, 1171, 1813, 23, 535, 665, 1434, 1439, 1184, 1695, 551, 815, 49, 438, 823, 312, 1738, 1745, 96, 1512, 635, 747, 1784, 1915}


#### 4. Nhập vào một từ điển (giả sử các values là các số nguyên). In ra các value có giá trị khác nhau.

In [10]:
d = {}
n = int(input("Nhập số phần tử từ điển: "))
for i in range(n):
    k = input("Nhập key: ")
    v = int(input("Nhập value: "))
    d[k] = v

print("Các value khác nhau:", set(d.values()))


Nhập số phần tử từ điển: 3
Nhập key: 2
Nhập value: 34
Nhập key: 34
Nhập value: 3
Nhập key: 2
Nhập value: 34
Các value khác nhau: {34, 3}


#### 5. Nhập vào một từ điển (giả sử các values là các số nguyên). In ra value có giá trị lớn nhất.

In [12]:
d = {}
n = int(input("Nhập số phần tử từ điển: "))
for i in range(n):
    k = input("Nhập key: ")
    v = int(input("Nhập value: "))
    d[k] = v

max_value = max(d.values())
print("Value lớn nhất:", max_value)


Nhập số phần tử từ điển: 3
Nhập key: 1
Nhập value: 2
Nhập key: 3
Nhập value: 4
Nhập key: 5
Nhập value: 6
Value lớn nhất: 6


#### 6. Nhập vào một từ điển (giả sử các keys là các số nguyên). Tìm hai khóa có giá trị lớn nhất, lớn nhì.

In [13]:
d = {}
n = int(input("Nhập số phần tử từ điển: "))
for i in range(n):
    k = int(input("Nhập key (số nguyên): "))
    v = int(input("Nhập value: "))
    d[k] = v

sorted_keys = sorted(d.keys(), reverse=True)
print("Khóa lớn nhất:", sorted_keys[0])
print("Khóa lớn nhì:", sorted_keys[1])


Nhập số phần tử từ điển: 3
Nhập key (số nguyên): 1
Nhập value: 2
Nhập key (số nguyên): 3
Nhập value: 4
Nhập key (số nguyên): 5
Nhập value: 6
Khóa lớn nhất: 5
Khóa lớn nhì: 3


### BT Bổ sung

#### 1. Cho biết kết quả in ra màn hình khi thực hiện các lệnh sau đây:


In [16]:
def A(list1):
    list1.append(2*list1[0])
    print(list1)
def B(list2):
    list2.append(2+list2[1])
    list2 = []
    print(list2)
def C(list1,list2):
    A(list1)
    print(list1)
    print(list2)
    B(list2)
    print(list2)

a = [0,1]
b = [2,3]
C(a,b)

[0, 1, 0]
[0, 1, 0]
[2, 3]
[]
[2, 3, 5]


### Phân tích 

* Ban đầu: a = [0,1], b = [2,3]
* C(a,b):

#### Gọi A(a) → list1.append(2*list1[0]) = 2*0 = 0 → a = [0,1,0]
#### In ra [0, 1, 0]

#### print(list1) → [0,1,0]

#### print(list2) → [2,3]

#### B(b):

* b.append(2+3)=5 → b = [2,3,5]

* Sau đó list2 = [] (gán cục bộ) → in ra []

#### print(list2) → vẫn là [2,3,5] (biến gốc b không bị mất)

#### 2. Cho biết kết quả in ra màn hình khi thực hiện các lệnh sau đây:

In [17]:
def myfunc(n): 
    M = list(range(n,0,-1))
    P = M
    print(M)
    M[0] = n+1
    print(P)
    Q = P + [10]
    Q.sort()
    print(Q)
    P.pop()
    print(P)
    return [P, M, Q]

print(myfunc(2))


[2, 1]
[3, 1]
[1, 3, 10]
[3]
[[3], [3], [1, 3, 10]]


#### Phân tích
* n=2 → M = [2,1]. In [2,1]
* P = M (cùng tham chiếu)
* M[0] = 3 → M = [3,1], nên P = [3,1]. In [3,1]
* Q = P + [10] = [3,1,10] → sort → [1,3,10]. In [1,3,10]
* P.pop() → xóa phần tử cuối 1 → P = [3] → đồng thời M = [3]. In [3]
* return [P,M,Q] = [[3], [3], [1,3,10]]

#### 1) Cho một list chứa các tuple (mỗi tuple chứa 3 thành phần (tên loại trái cây, giá tiền, số lượng trái cây), trong đó tên loại trái cây là một chuỗi là các chữ in thường, giá tiền là số thực tính theo đơn vị $, số lượng trái cây là số nguyên dương)
##### Ví dụ: input list như sau: L = [('lemon', 0.45, 15), ('apple', 2.5, 3), ('peach', 4.3, 10), ('banana', 0.6, 5)]
* a) Viết chương trình tạo ra một list chứa tên của tất cả các loại trái cây trong L thỏa điều kiện các tên đều là chữ in hoa và được sắp xếp tăng theo thứ tự từ điển
##### Ví dụ: output như sau: ['APPLE', 'BANANA', 'LEMON', 'PEACH']
* b) Viết chương trình tính tổng số lượng tất cả các loại trái cây có trong list đã cho
##### Ví dụ: output là 15+3+10+5 bằng 33
* c) Viết chương trình sắp xếp list của các tuple theo thứ tự giảm dần về số lượng trái cây trong tuple
##### Ví dụ: output là [('lemon', 0.45, 15), ('peach', 4.3, 10), ('banana', 0.6, 5), ('apple', 2.5, 3)]

#### 2) Cho một list chứa các chuỗi (mỗi chuỗi gồm 3 thành phần cách nhau bởi dấu phẩy có dạng như sau: "<Tên Khoa>, <Khóa học>, <Số lượng sinh viên>", trong đó Tên Khoa là thông tin của một Khoa biễu diễn dạng chuỗi, Khóa học là thông tin một khóa học biễu diễn dạng số nguyên dương trong chuỗi, Số lượng sinh viên là biễu diễn dạng số nguyên dương trong chuỗi). Giả sử rằng Tên Khoa không xuất hiện trùng trong các chuỗi trong list
Ví dụ input list L như sau: L = ["CNTT,20,600", "Ly,18,200", "Toan,19,100", "Hoa,19,150"]
* a) Viết chương trình tạo ra một từ điển trong đó các khóa (key, dạng chuỗi) của từ điển là Tên Khoa của các chuỗi trong list đã cho; các giá trị (value, dạng số nguyên) tương ứng là số lượng sinh viên tương ứng của mỗi Khoa
Ví dụ: ouput từ điển D là {'CNTT': 600, 'Ly': 200, 'Toan': 100,'Hoa': 150}
* b) Viết chương trình tính tổng tất cả các giá trị (value) trong từ điển của câu (a)
Ví dụ: tổng là 1050
* c) In ra khóa trong từ điển tìm được ở câu (a) có giá trị (value) lớn nhất.
Ví dụ: khóa với giá trị lớn nhât là CNTT